src/storage/compactor/src/server.rs

```diff
@@ -272,3 +279,261 @@ pub async fn compactor_serve(
+        sstable_size_mb * (1 << 20),
+        block_size_kb * (1 << 10),
+    );
+
+    // check memory config
+    {
+        // This is a similar logic to SstableBuilder memory detection, to ensure that we can find
+        // configuration problems as quickly as possible
+        let min_compactor_memory_limit_bytes =
+            (sstable_size_mb * (1 << 20) + block_size_kb * (1 << 10)) as u64;
+
+        assert!(compactor_memory_limit_bytes > min_compactor_memory_limit_bytes * 2);
+    }
+
+    let mut object_store = parse_remote_object_store_with_config(
+        state_store_url
+            .strip_prefix("hummock+")
+            .expect("object store must be hummock for compactor server"),
+        object_metrics,
+        "Hummock",
+        None,
+    )
+    .await;
+    object_store.set_opts(
+        object_store_streaming_read_timeout_ms,
+        object_store_streaming_upload_timeout_ms,
+        object_store_read_timeout_ms,
+        object_store_upload_timeout_ms,
+    );
+    let object_store = Arc::new(object_store);
+    let sstable_store = Arc::new(SstableStore::for_compactor(
+        object_store,
+        data_directory,
+        1 << 20, // set 1MB memory to avoid panic.
+        meta_cache_capacity_bytes,
+    ));
+
+    // let telemetry_enabled = system_params_reader.telemetry_enabled();
+
+    // let filter_key_extractor_manager = Arc::new(FilterKeyExtractorManager::new(Box::new(
+    //     RemoteTableAccessor::new(meta_client.clone()),
+    // )));
+    // let system_params_manager = Arc::new(LocalSystemParamsManager::new(system_params_reader));
+    // let compactor_observer_node = CompactorObserverNode::new(
+    //     filter_key_extractor_manager.clone(),
+    //     system_params_manager.clone(),
+    // );
+    // let observer_manager =
+    //     ObserverManager::new_with_meta_client(meta_client.clone(),
+    // compactor_observer_node).await;
+
+    // use half of limit because any memory which would hold in meta-cache will be allocate by
+    // limited at first.
+    // let observer_join_handle = observer_manager.start().await;
+
+    let memory_limiter = Arc::new(MemoryLimiter::new(compactor_memory_limit_bytes));
+    let memory_collector = Arc::new(CompactorMemoryCollector::new(
+        sstable_store.clone(),
+        memory_limiter.clone(),
+        storage_memory_config,
+    ));
+
+    monitor_cache(memory_collector, &registry).unwrap();
+
+    let await_tree_config = match &config.streaming.async_stack_trace {
+        AsyncStackTraceOption::Off => None,
+        c => await_tree::ConfigBuilder::default()
+            .verbose(c.is_verbose().unwrap())
+            .build()
+            .ok(),
+    };
+    let await_tree_reg =
+        await_tree_config.map(|c| Arc::new(RwLock::new(await_tree::Registry::new(c))));
+
+    // The following will be passed via DispatchCompactionTaskRequest, so here is just a simulation.
```

Comment:
WIP,  `output_ids`, `id_to_table`, `dispatch_task` will be changed to receive RPC to get.